File: sdk/typescript/src/reports.ts
import type { ReconciliationReport } from "./models.js";

function decodeBase64Url(value: string): Uint8Array {
  const padded = value.replaceAll("-", "+").replaceAll("_", "/") + "=".repeat((4 - value.length % 4) % 4);
  return Uint8Array.from(atob(padded), (character) => character.charCodeAt(0));
}
function hex(bytes: Uint8Array): string { return [...bytes].map((value) => value.toString(16).padStart(2, "0")).join(""); }
function ownedBuffer(bytes: Uint8Array): ArrayBuffer { const copy = new Uint8Array(bytes.byteLength); copy.set(bytes); return copy.buffer; }

/** Verify downloaded bytes and the detached report signature against the key ID frozen on the report. */
export async function verifyReconciliationReport(bytes: Uint8Array, report: ReconciliationReport, publicKeys: Readonly<Record<string, Uint8Array>>): Promise<void> {
  if (report.status !== "ready" || !report.object_sha256 || !report.signature || !report.signing_key_id) throw new Error("report is not ready or lacks integrity metadata");
  const publicKey = publicKeys[report.signing_key_id];
  if (!publicKey) throw new Error(`unknown reconciliation signing key: ${report.signing_key_id}`);
  const digest = new Uint8Array(await crypto.subtle.digest("SHA-256", ownedBuffer(bytes)));
  if (hex(digest) !== report.object_sha256) throw new Error("reconciliation report digest mismatch");
  if (report.object_size_bytes !== undefined && BigInt(report.object_size_bytes) !== BigInt(bytes.byteLength)) throw new Error("reconciliation report size mismatch");
  const prefix = new TextEncoder().encode(`merchant-reconciliation-jsonl-v1\0${report.id}\0${report.snapshot_ledger_sequence}\0`);
  const message = new Uint8Array(prefix.length + digest.length); message.set(prefix); message.set(digest, prefix.length);
  const key = await crypto.subtle.importKey("raw", ownedBuffer(publicKey), { name: "Ed25519" }, false, ["verify"]);
  if (!await crypto.subtle.verify("Ed25519", key, ownedBuffer(decodeBase64Url(report.signature)), ownedBuffer(message))) throw new Error("reconciliation report signature mismatch");
}

File: sdk/go/reports.go
package merchantplatform

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"strconv"
)

func ReconciliationSignatureMessage(reportID, snapshotSequence string, digest []byte) []byte {
	message := append([]byte("merchant-reconciliation-jsonl-v1\x00"), []byte(reportID)...)
	message = append(message, 0)
	message = append(message, []byte(snapshotSequence)...)
	message = append(message, 0)
	return append(message, digest...)
}

// VerifyReconciliationReport consumes raw, checking exact bytes, size and the key ID frozen on the ready report.
func VerifyReconciliationReport(raw io.Reader, report ReconciliationReport, publicKeys map[string]ed25519.PublicKey) error {
	if report.Status != "ready" {
		return errors.New("report is not ready")
	}
	hash := sha256.New()
	size, err := io.Copy(hash, raw)
	if err != nil {
		return err
	}
	digest := hash.Sum(nil)
	if hex.EncodeToString(digest) != report.ObjectSHA256 {
		return errors.New("reconciliation report digest mismatch")
	}
	if report.ObjectSizeBytes != "" {
		expected, err := strconv.ParseInt(report.ObjectSizeBytes, 10, 64)
		if err != nil || expected != size {
			return errors.New("reconciliation report size mismatch")
		}
	}
	key, ok := publicKeys[report.SigningKeyID]
	if !ok {
		return fmt.Errorf("unknown reconciliation signing key: %s", report.SigningKeyID)
	}
	signature, err := base64.RawURLEncoding.DecodeString(report.Signature)
	if err != nil || !ed25519.Verify(key, ReconciliationSignatureMessage(report.ID, report.SnapshotLedgerSequence, digest), signature) {
		return errors.New("reconciliation report signature mismatch")
	}
	return nil
}
